src/Microsoft.AspNetCore.Blazor.Razor.Extensions/InjectDirective.cs

```diff
@@ -0,0 +1,114 @@
+{
+    // Much of the following is equivalent to Microsoft.AspNetCore.Mvc.Razor.Extensions's InjectDirective,
+    // but this one outputs properties annotated for Blazor's property injector, plus it doesn't need to
+    // support multiple CodeTargets.
+
+    internal class InjectDirective
+    {
+        const string InjectAttributeTypeName = "Microsoft.AspNetCore.Blazor.Components.InjectAttribute";
+
+        public static readonly DirectiveDescriptor Directive = DirectiveDescriptor.CreateDirective(
+            "inject",
+            DirectiveKind.SingleLine,
+            builder =>
+            {
+                builder.AddTypeToken("TypeName", "The type of the service to inject.");
+                builder.AddMemberToken("PropertyName", "The name of the property.");
+                builder.Usage = DirectiveUsage.FileScopedMultipleOccurring;
+                builder.Description = "Inject a service from the application's service container into a property.";
+            });
+
+        public static void Register(IRazorEngineBuilder builder)
+        {
+            builder.AddDirective(Directive);
+            builder.Features.Add(new Pass());
+        }
+
+        private class Pass : IntermediateNodePassBase, IRazorDirectiveClassifierPass
+        {
+            protected override void ExecuteCore(
+                RazorCodeDocument codeDocument,
+                DocumentIntermediateNode documentNode)
+            {
+                var visitor = new Visitor();
+                visitor.Visit(documentNode);
+
+                var properties = new HashSet<string>(StringComparer.Ordinal);
+                var classNode = documentNode.FindPrimaryClass();
+
+                for (var i = visitor.Directives.Count - 1; i >= 0; i--)
+                {
+                    var directive = visitor.Directives[i];
+                    var tokens = directive.Tokens.ToArray();
+                    if (tokens.Length < 2)
+                    {
+                        continue;
+                    }
+
+                    var typeName = tokens[0].Content;
+                    var memberName = tokens[1].Content;
+
+                    if (!properties.Add(memberName))
+                    {
+                        continue;
+                    }
+
+                    classNode.Children.Add(new InjectIntermediateNode(typeName, memberName));
+                }
+            }
+
+            private class Visitor : IntermediateNodeWalker
+            {
+                public IList<DirectiveIntermediateNode> Directives { get; }
+                    = new List<DirectiveIntermediateNode>();
+
+                public override void VisitDirective(DirectiveIntermediateNode node)
+                {
+                    if (node.Directive == Directive)
+                    {
+                        Directives.Add(node);
+                    }
+                }
+            }
+
+            internal class InjectIntermediateNode : ExtensionIntermediateNode
```

Comment:
I think you already have a good intuition about this, but FYI:

The design intent of the IR is to provide nodes for all of the 'fundamental truths' of Razor. That includes, html, html attribute values (which have some special semantics), c# expressions, c# 'blobs', as well as generic c# structure like classes and namespaces. 

As such, all of these concepts in the IR are intentionally inflexible. What we define needs to work for everyone because they are fundamental to the language. You can implement these concepts in a different way (as you have done) by replacing the node writer.

------

Tag Helpers also fall into this category of 'fundamental truth' and represent the invocation syntax of an html tag that will **run code** to output html. They are intentionally abstracted from the runtime implementation, and can map to different implementations inside the framework (MVC already does this).

When we get further along (maybe tomorrow) - we'll have derived extension nodes that represent the specifics of components, and probably also a target extension that renders them to c#. Example - all of the VCTH stuff [here](https://github.com/aspnet/Razor/tree/dev/src/Microsoft.AspNetCore.Mvc.Razor.Extensions)

Using a target extension and defining lots of nodes is a little more ceremony, but provides some design benefits when doing something complex. In particular it's helpful when you need to separate editor codegen from runtime code gen.

-----

The third category of stuff is extension nodes that represent framework-specific concepts (such as `@page` from MVC).

It always a cost benefit choice of whether to implement a target extension for these things.  We made the tradeoff in terms of implementing a target extension for most cases, but I can't say it's always paid off. It can always be added later.

I was bored while making coffee 😆

Reply:
OK, thanks for the info!